File: python_mocra/mq/compression.py
import gzip
from typing import Union

try:
    import zstandard as zstd
except Exception:  # pragma: no cover - optional dependency
    zstd = None


ZSTD_MAGIC = b"\x28\xB5\x2F\xFD"
GZIP_MAGIC = b"\x1f\x8b"


def compress_payload(data: bytes, threshold: int = 1024) -> bytes:
    if len(data) <= threshold:
        return data
    if zstd is not None:
        try:
            return zstd.ZstdCompressor(level=3).compress(data)
        except Exception:
            pass
    try:
        return gzip.compress(data)
    except Exception:
        return data


def decompress_payload(payload: bytes) -> bytes:
    if len(payload) >= 4 and payload[:4] == ZSTD_MAGIC and zstd is not None:
        try:
            return zstd.ZstdDecompressor().decompress(payload)
        except Exception:
            return payload
    if len(payload) >= 2 and payload[:2] == GZIP_MAGIC:
        try:
            return gzip.decompress(payload)
        except Exception:
            return payload
    return payload

File: src/queue/compression.rs
use flate2::read::GzDecoder;
use std::borrow::Cow;
use std::io::Read;

pub fn compress_payload(data: &[u8], threshold: usize) -> Vec<u8> {
    if data.len() > threshold {
        match zstd::stream::encode_all(data, 3) {
            Ok(compressed) => compressed,
            Err(_) => data.to_vec(),
        }
    } else {
        data.to_vec()
    }
}

pub fn compress_payload_owned(data: Vec<u8>, threshold: usize) -> Vec<u8> {
    if data.len() > threshold {
        match zstd::stream::encode_all(data.as_slice(), 3) {
            Ok(compressed) => compressed,
            Err(_) => data,
        }
    } else {
        data
    }
}

pub fn decompress_payload<'a>(payload: &'a [u8]) -> Cow<'a, [u8]> {
    // Try Zstd first (magic: 0x28 0xB5 0x2F 0xFD)
    if payload.len() > 4
        && payload[0] == 0x28 && payload[1] == 0xB5
        && payload[2] == 0x2F && payload[3] == 0xFD
    {
        match zstd::stream::decode_all(payload) {
            Ok(decoded) => Cow::Owned(decoded),
            Err(_) => Cow::Borrowed(payload),
        }
    } else if payload.len() > 2 && payload[0] == 0x1f && payload[1] == 0x8b {
        // Fallback to Gzip
        let mut decoder = GzDecoder::new(payload);
        let mut s = Vec::new();
        if decoder.read_to_end(&mut s).is_ok() {
            Cow::Owned(s)
        } else {
            Cow::Borrowed(payload)
        }
    } else {
        Cow::Borrowed(payload)
    }
}
